lua/diffview/vcs/adapters/hg/init.lua

```diff
@@ -1,18 +1,1141 @@
+  cfile = pl:readlink(cfile) or cfile
+
+  for _, path in ipairs(paths) do
+    table.insert(top_indicators, pl:absolute(path, cpath))
+    break
+  end
+
+  table.insert(top_indicators, cpath and pl:realpath(cpath) or (
+    vim.bo.buftype == ""
+    and pl:absolute(cfile)
+    or nil
+  ))
+
+  if not cpath then
+    table.insert(top_indicators, pl:realpath("."))
+  end
+
+  return paths, top_indicators
+end
+
+---Get the git toplevel directory from a path to file or directory
+---@param path string
+---@return string?
+local function get_toplevel(path)
+  local out, code = utils.system_list(vim.tbl_flatten({config.get_config().hg_cmd, {"root"}}), path)
+  if code ~= 0 then
+    return nil
+  end
+  return out[1] and vim.trim(out[1])
 end
 
 function M.find_toplevel(top_indicators)
-  -- TODO: implement
-  return "", nil
+  local toplevel
+
+  for _, p in ipairs(top_indicators) do
+    if not pl:is_dir(p) then
+      p = pl:parent(p)
+    end
+
+    if p and pl:readable(p) then
+      toplevel = get_toplevel(p)
+      if toplevel then
+        return nil, toplevel
+      end
+    end
+  end
+
+  return (
+    ("Path not a mercurial repo (or any parent): %s")
+    :format(table.concat(vim.tbl_map(function(v)
+      local rel_path = pl:relative(v, ".")
+      return utils.str_quote(rel_path == "" and "." or rel_path)
+    end, top_indicators) --[[@as vector ]], ", "))
+  ), nil
+end
+
+function M.create(toplevel, path_args, cpath)
+  return HgAdapter({
+    toplevel = toplevel,
+    path_args = path_args,
+    cpath = cpath,
+  })
+end
+
+function HgAdapter:init(opt)
+  opt = opt or {}
+  HgAdapter:super().init(self, opt)
+
+  self.ctx = {
+    toplevel = opt.toplevel,
+    dir = opt.toplevel,
+    path_args = opt.path_args or {},
+  }
+
+  self:init_completion()
+end
+
+function HgAdapter:get_command()
+  return config.get_config().hg_cmd
+end
+
+function HgAdapter:get_show_args(path, rev)
+  return utils.vec_join(self:args(), "cat", "--rev", rev:object_name(), "--", path)
+end
+
+function HgAdapter:get_log_args(args)
+  return utils.vec_join("log", "--stat", '--rev', args)
+end
+
+function HgAdapter:get_merge_context()
+  local ret = {}
+
+  local out, code = self:exec_sync({ "debugmergestate", "-Tjson" }, self.ctx.toplevel)
+
+  if code ~= 0 then
+    return {}
+  end
+
+  local data = vim.json.decode(table.concat(out, ""))
+  for _, commit in ipairs(data[1].commits) do
+    if commit.name == "other" then
+      ret.theirs = { hash = commit.node }
+      out, code = self:exec_sync({ "log", "--template={branch}", "--rev", commit.node })
+      if code == 0 then
+        ret.theirs.ref_name = out[1]
+      end
+    elseif commit.name == "local" then
+      ret.ours = { hash = commit.node }
+      out, code = self:exec_sync({ "log", "--template={branch}", "--rev", commit.node })
+      if code == 0 then
+        ret.ours.ref_name = out[1]
+      end
+    end
+  end
+
+  -- Base is ancestorlinknode (per file)
+  ret.base = { hash = "" }
```

Comment:
It seems hg gives me the merge-base per file, which is not always the same.
So to be able to give that commit, I would need to know which file is being viewed currently. Maybe something for a follow up?

Reply:
Really? The merge base can be different for different files within a single conflict? How does that work? Are we confusing terminology here perhaps? In git the merge base refers to the common ancestor of all the branches involved in a merge. Thus, by definition, there can only be one.

But sure, we can address this in a follow up PR. Add a `TODO` comment here or something so that we don't forget to come back to this :)

Reply:
I created a little test case with 2 files, and I do get a merge-base revision per file (it's called `ancestorlinknode`) but it is indeed the same for all files involved, except if the file is new (then it is `00000000`).

The testcase I was using before was a bit of a stretch, I had created the same new file in both branches.

So I guess I can use the first non-zero `ancestorlinknode` and assume it is the same for all.